### PR TITLE
Feature/user settings

### PR DESCRIPTION
**Pull request by Ravikant96**
Description:
Describe your pull request and attach links to any issues that may be related.

### CODE REVIEW

File: server/www/settings.js [hunk 1]
```diff
@@ -1,16 +1,19 @@
 const API = require("../utils/api");
 const account = require('../onServerStart');
 const commonFun = require("../utils/commonFunctions");
+const redis = require("../utils/redis").Redis;
 
 exports.insert = class extends API {
 
-	async insert({profile, owner, value, account_id} = {}) {
+	async insert({profile, owner, value, owner_id} = {}) {
```
Comment (Ravikant96) [inline server/www/settings.js]: owner id ka check for user case. admin surperadmin ya own

File: server/www/settings.js [hunk 2]
```diff
@@ -20,43 +23,49 @@ exports.insert = class extends API {
 						account_id,
 						profile,
 						owner,
+						owner_id,
 						value
 					)
 				VALUES
```
Comment (Ravikant96) [inline server/www/settings.js]: one less

File: server/www/settings.js [hunk 3]
```diff
@@ -20,43 +23,49 @@ exports.insert = class extends API {
...
+		}
+
+		if(owner == 'user') {
+			await redis.del(`user.settings.${this.user.user_id}`);
+		}
 
 		this.assert(id, "no id found to update");
 		this.assert(commonFun.isJson(value), "Please send valid JSON");
 
 		await account.loadAccounts();
```
Comment (Ravikant96) [inline server/www/settings.js]: shift it after query exec

File: server/www/users.js [hunk 4]
```diff
@@ -518,6 +521,10 @@ exports.metadata = class extends API {
 		metadata.visualizations = visualizations;
 		metadata.datasources = datasources;
 
+		this.assert(commonFun.isJson(userSettings[0].value), 'userSettings.value is not a json');
```
Comment (Ravikant96) [inline server/www/users.js]: length

File: server/utils/User.js [hunk 5]
```diff
@@ -9,9 +12,40 @@ class User {
...
+		const userSetting = await redis.hget(`user.settings.${this.user_id}`, profile);
+
+		if(userSetting) {
+
+			this.assert(commonFun.isJson(userSetting), 'UserSettings is not a JSON')
+
+			return JSON.parse(userSetting);
+		}
+
+		const response = mysql.query(`SELECT * FROM tb_settings WHERE status = 1 AND owner = ? AND owner_id = ?`, ['user', this.user_id]);
```
Comment (Ravikant96) [inline server/utils/User.js]: await

Comment (Ravikant96) [reply server/utils/User.js]: Ownerid join with tb_user to check if user hai ya nai

File: server/utils/User.js [hunk 6]
```diff
@@ -9,9 +12,40 @@ class User {
...
+		Object.assign(this, userObj);
+	}
+
+	async get(profile) {
+
+		const userSetting = await redis.hget(`user.settings.${this.user_id}`, profile);
+
+		if(userSetting) {
+
+			this.assert(commonFun.isJson(userSetting), 'UserSettings is not a JSON')
```
Comment (Ravikant96) [inline server/utils/User.js]: try catch.... return status and error and message

File: client/js/main.js [hunk 7]
```diff
@@ -524,6 +525,17 @@ Page.serviceWorker = class PageServiceWorker {
 	get status() {
 		return navigator.serviceWorker.controller ? true : false;
 	}
+
+	async clear() {
+
+		if(!this.status)
+			return false;
+
+		if(navigator.serviceWorker) {
```
Comment (achshar) [inline client/js/main.js]: Not needed.

File: client/js/user/settings.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		this.page = page;
+	}
+
+	async setup() {
+
+		const settings_json = [
+			{
+				key: 'theme',
+				type: 'multiselect',
+				name: 'Themes',
```
Comment (achshar) [inline client/js/user/settings.js]: Theme

File: client/js/user/settings.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+				description: 'Themes for user',
```
Comment (achshar) [inline client/js/user/settings.js]: Not needed.

File: client/js/user/settings.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+		];
+
+		this.settingsManger = new SettingsManager({owner: 'user', owner_id: user.user_id, format: settings_json, disable_aside: false});
+
+		await this.settingsManger.load();
+
+		this.render();
+	}
+
+	render() {
```
Comment (achshar) [inline client/js/user/settings.js]: Not needed

File: client/index.js [hunk 11]
```diff
@@ -32,10 +31,27 @@ class HTMLAPI extends API {
...
+		if(await this.account.settings.get('theme'))
+			theme = await this.account.settings.get('theme');
 
-		if(this.account.settings.get('custom_css'))
+
+		if(this.user && await this.user.settings.get('theme'))
+			theme = await this.user.settings.get('theme');
+
+
+		if(theme)
```
Comment (achshar) [inline client/index.js]: Not needed

File: client/js/user/settings.js [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+				datalist: [
+					{name: 'Dark', value: 'dark'},
+					{name: 'Light', value: 'light'},
+				],
+				multiple: false,
+				dropDownPosition: 'bottom',
+			},
+		];
+
+		const settingsManger = new SettingsManager({owner: 'user', owner_id: user.user_id, format: settings_json, disable_aside: false});
```
Comment (achshar) [inline client/js/user/settings.js]: Open

**Pull request merged**